pkg/rpm/preupgrade

```diff
@@ -0,0 +1,356 @@
+    return data
+
+
+def get_tpool_stats(config):
+    """
+    Returns a mapping representing data for the serviced thinpool.
+    """
+    thinpooldev = config.get("SERVICED_DM_THINPOOLDEV", "serviced")
+    cmd = "lvs -o lv_size,data_percent,lv_metadata_size,metadata_percent %s 2>/dev/null | grep -vi lsize" % thinpooldev
+    stats = subprocess.check_output(cmd, shell=True).strip()
+    if not stats:
+        return {}
+    data = calc_tpool_stats(stats)
+    data['thinpooldev'] = thinpooldev
+    return data
+
+
+def get_tenant_stats():
+    """
+    Uses df to get free space for mounted tenant volumes (if any). df -h shows
+    base 1024 stats, so we need to parse these appropriately.
+    """
+    data = { 'tenants': [] }
+    cmd = "df -h --output=avail,target | grep /opt/serviced/var/volumes 2>/dev/null || true"
+    stats = subprocess.check_output(cmd, shell=True)
+    if not stats:
+        return data
+    for line in [s.strip() for s in stats.split('\n') if s != '']:
+        line = line.split(' ')
+        if len(line) != 2: continue
+        tenant = line[1].split('/')[-1]
+        free = parse_size(line[0].lower())
+        data['tenants'].append({ 'id': tenant, 'free': free })
+    return data
+
+
+def get_thinpool_data():
+    """
+    Returns a map of data for the serviced thin pool.
+    """
+    config = parse_serviced_config()
+    data = get_serviced_settings(config)
+    data.update(get_tpool_stats(config))
+    data.update(get_tenant_stats())
+    return data
+
+
+def app_is_deployed():
+    """
+    Returns True if an application has been deployed, otherwise False
+    """
+    result = subprocess.check_output("serviced service list 2>/dev/null || true", shell=True).strip()
+    return len(result) > 0
+
+
+# Returns "" if there are no problems, otherwise returns a string explaining
+# what was too small.
+def check_thinpool(stats):
+    err = ""
+    if stats['data_free'] < stats['data_min_free']:
+        err += "\nError: The thinpool storage free (%s) is under the minimum threshold (%s)" % \
+            (bytesize(stats['data_free']), bytesize(stats['data_min_free']))
+    if stats['meta_free'] < stats['meta_min_free']:
+        err += "\nError: The thinpool metadata free (%s) is under the minimum threshold (%s)" % \
+            (bytesize(stats['meta_free']), bytesize(stats['meta_min_free']))
+        err += "\n%sThe metadata minimum threshold is calculated as 2 percent of the SERVICED_STORAGE_MIN_FREE value.%s" % \
+            (C_RESET, C_ERR)
+    if len(stats['tenants']):
+        for tenant in stats['tenants']:
+            if tenant['free'] < stats['data_min_free']:
+                err += "\nError: The tenant volume %s available space (%s) is under the minimum threshold (%s)" % \
+                    (tenant['id'], bytesize(tenant['free']), bytesize(stats['data_min_free']))
+    else:
+        # Only return an error here if an app has been deployed.
+        if app_is_deployed():
```

Comment:
This check is problematic for a couple of reasons. First, the check assumes that serviced is running during the upgrade; the user should be able to upgrade CC even if it's stopped.   

If the machine has just been rebooted and serviced hasn't been started, then no DM device will be active, and nothing will be mounted.  The same is true in the case of HA on the secondary (passive) node - there will be no DM device active so nothing will be mounted.